File: Striver Important Interview Questions/Day 21 - Binary Search Tree Part 2/kth-smallest-element-in-a-bst.cpp

// https://leetcode.com/problems/kth-smallest-element-in-a-bst/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:

    void solve(TreeNode*root, int k, int &i, int &ans){
        if(!root) return;

        solve(root->left, k, i, ans);
        i+=1;
        if(i == k){
            ans = root->val;
            return;
        }

        solve(root->right, k, i, ans);
        return;
    }

    int kthSmallest(TreeNode* root, int k) {
        int ans = -1;
        if(!root) return ans;

        int i = 0;
        solve(root, k, i, ans);
        
        return ans;
    }
};

File: Neetcode-150-js/Trees/kth-smallest-element-in-a-bst.js

// https://leetcode.com/problems/kth-smallest-element-in-a-bst/

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {
    let ans = -1;
    if(!root) return ans;
    let count = 0;
    
    var solve = (root) => {
        if(!root) return;

        // if(count === k){
        //     ans = root.val;
        //     return;
        // }
        solve(root.left);
        count++;
        if(count === k){
            ans = root.val;
            return;
        }
        solve(root.right);
        return;
    }

    solve(root);
    return ans;
};